Unit 3.3 webpack

nathanjophillips opened the pull request:
## The Purpose
- Gain basic understand of webpack and how to incorporate it into this project
- Setup project file structure so that it can continue to be build upon in an organized way
- Refactor npm scripts to use Webpack

## What It Does
1. This adds necessary dependencies to properly run webpack in this project.
2. This changes the current npm `start` script to run `webpack-dev-server` and adds a `build` command as well.
3. This copies in the code and file structure from the given [Webpack-boilerplate](https://github.com/Banno/ux-intern-onboarding/tree/master/resources/webpack/webpack-boilerplate) to be used throughout the rest of this project.
4. With all this added the files will be hosted on [http://localhost:1820/webpack-dev-server](http://localhost:1820/webpack-dev-server) to display the bundled files and the static `html` files from the previous branch

## This accomplishes the following requirements given for  [Unit 3.3](https://github.com/phillnab/nathan-phillips-ux-iop/blob/master/issues/unit-3/3.3-webpack.md)

- You've added new dev dependencies:
  - [x] `webpack`
  - [x] `webpack-cli`
  - [x] `webpack-dev-server`
  - [x] `babel-core`
  - [x] `babel-loader`
  - [x] `polymer-webpack-loader`
  - [x] `file-loader`
  - [x] `css-loader`
  - [x] `html-webpack-plugin`
  - [x] `copy-webpack-plugin`

- You've added the Banno-scoped Polymer and design components as normal dependencies:
  - [x] `@banno/polymer`
  - [x] `@banno/jha-design-components`
- [x] `webpack.config.js` file exists in the root of the project
- [x] You've added the `jha-design-components-loader` tool in the `tools` directory
- [x] `webpack` command returns no errors
- [x] The `start` script now runs `webpack-dev-server`
- [x] A new script has been added to `package.json` called `build` which simply runs `webpack`
- [x] You've copied over all content from the `webpack-boilerplate` directory in the project resources to your project's `src` directory

### The changes to look for are:
1. Check for errors or warnings in the terminal 
2. Check for syntax mistakes in the `webpack.config.js` and `package.json` files
3. If it doesn't work in chrome on it's own check if it works in an incognito window
4. Once running check if each page is showing with css applied with icons


# How To Test
- Clone Repo [nathan-phillips-ux-iop](https://github.com/phillnab/nathan-phillips-ux-iop)
- Checkout branch `nathan-phillips-ux-iop/unit-3.3-webpack`
- Run `yarn`
- Run `yarn build`
- Run `yarn start`

## Step 1 Browser
- Once `yarn start` is completed copy in `http://localhost:1820/webpack-dev-server ` to the browser and see if this matches the screenshot.

- If it's not showing try it in an incognito window.

<img width="1256" alt="Screen Shot 2019-04-08 at 4 12 02 PM" src="https://user-images.githubusercontent.com/29068248/55757301-37e5cc80-5a19-11e9-9167-fd7e9b88bbba.png">


## Step 2 Nav through html files
- Enter the `index.html` file within the `html` folder and click through the menu bar to see if each of these pages show up like the following.
<img width="674" alt="Screen Shot 2019-04-08 at 4 12 14 PM" src="https://user-images.githubusercontent.com/29068248/55757461-93b05580-5a19-11e9-89c6-a7ea82c64875.png">
<img width="709" alt="Screen Shot 2019-04-08 at 4 12 20 PM" src="https://user-images.githubusercontent.com/29068248/55757467-98750980-5a19-11e9-854c-f2aadb4461f9.png">
<img width="741" alt="Screen Shot 2019-04-08 at 4 12 25 PM" src="https://user-images.githubusercontent.com/29068248/55757475-9a3ecd00-5a19-11e9-8692-f09d3db2f2f8.png">
<img width="713" alt="Screen Shot 2019-04-08 at 4 12 31 PM" src="https://user-images.githubusercontent.com/29068248/55757480-9ca12700-5a19-11e9-9c02-75e4a82d71e7.png">


## Step 3 file structure
- Open your code editor and check to see if the files follow this model
<img width="445" alt="Screen Shot 2019-04-02 at 5 08 33 PM" src="https://user-images.githubusercontent.com/29068248/55747932-0104bc00-5a03-11e9-9e68-289744087c42.png">


## Checklist

Under penalty of public shaming, I avow that I:

- [x] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [x] Verified that there are no compiler or linter errors or warnings
- [ ] Verified the build in production(optimized) mode
- [x] Updated the docs, if necessary
- [x] Included the appropriate labels
- [x] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] Internet Explorer

Commit:
Added webpack-boilerplate
--devDependencies
--new files structure
--webpack.config.js file
--new scripts in package.json to build and run on webpack-dev-server

Commit:
fixed babel issue

Commit:
fixed build errors

Commit:
changed path in index.html to stylesheet

Commit:
Changed polymerElement to polymer.Element

Commit:
added settings to .gitignore

Jropp commented on .gitignore:
```diff
@@ -1 +1,2 @@
 node_modules
+.vscode/settings.json
```
You could add your dist folder to this so it doesn't try to upload your build files

kddrake replied:
should, not could

Jropp commented on webpack.config.js:
```diff
@@ -0,0 +1,81 @@
+					]
+        },
+				{
+          test: /polymer\.html$/,
```
So I don't think this is actually doing anything as it doesn't have a loader on it. I'd comment it out and test whether or not it is actually necessary

Jropp commented on webpack.config.js:
```diff
@@ -0,0 +1,81 @@
+			contentBase: path.join(__dirname, 'dist'),
+			compress: true,
+			port: 1820
+
```
unneeded space

Jropp commented on webpack.config.js:
```diff
@@ -0,0 +1,81 @@
+    },
+    devtool: 'source-map',
+		plugins: [
+      new webpack.NormalModuleReplacementPlugin(
```
This doesn't need changed, I just wanted to point out here that this is why the icons importing polymer.html instead of polymer-element.html isn't a problem. This is changing that path so it actually is requiring polymer-element.html

Jropp commented on dist/bundle.js:
```diff
@@ -0,0 +1,2 @@
+!function(t){var e={};function r(i){if(e[i])return e[i].exports;var s=e[i]={i:i,l:!1,exports:{}};return t[i].call(s.exports,s,s.exports,r),s.l=!0,s.exports}r.m=t,r.c=e,r.d=function(t,e,i){r.o(t,e)||Object.defineProperty(t,e,{enumerable:!0,get:i})},r.r=function(t){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(t,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(t,"_ ...
```
Delete the content of the dist folder and add the folder to the .gitignore

Jropp commented on index.html:
```diff
@@ -3,7 +3,7 @@
   <meta charset="UTF-8"/>
   <title>User Profile</title>
-  <link rel="stylesheet" type="text/css" href="style.css" />
+  <link rel="stylesheet" type="text/css" href="'dist/style.css'" />
```
Get rid of extra quotes here.

But also, is there a reason you are importhing this here from your build folder? This isn't the index.html that will be bundled and served.

Jropp commented on index.html:
```diff
@@ -63,7 +63,7 @@ <h3>User Profile</h3>
       </tr>
     </table>
-    <script>
+    <script type="text/javascript" src="dist/bundle.js">
```
Also here, is there a specific reason you are bringing in the app bundle as this index.html is not what is getting served up by webpack?

Jropp reviewed (approved):
It is working fine for me without any errors. 

On top of these changes you could also delete the custom loader that is in the tools file, as you won't need it for the jha icons anymore

Commit:
copied html files into dist foler to be hosted

kddrake reviewed (changes requested):
- Can you remove the starter-component stuff? you won't actually ever need it, other than maybe to reference as you build out your own components.
- Pull `index.html` out of the `html`, and let it live in `src/`. Then when you're having people check the app, you _should_ just need to navigate them to `http://localhost:1820/index.html` rather than having them route through the dev server thing.

Commit:
Final cleanup

-- added dist to gitignore
-- removed starter components
-- removed polymer .html test in webpack.config
